### ISSUE TITLE
Ability to install pinned deps

### ISSUE THREAD

**Issue by maraisr** [volta-cli/volta#653]
Is there a command I can run that'll install tools that are pinned? Without first running them?

ie, i have this in my `package.json`, but I have node@12 installed.

```json
{
  "volta": {
    "node": "13.7.0",
    "yarn": "1.21.1"
  }
}
```

without first running `node`, just to install them. Id like to run something like `yarn install pinned` or something.

**Comment by charlespierce** [comment]
Hi @maraisr, we don't currently have anything that will only fetch, but not run, the pinned values from `package.json`. We do have `volta fetch`, but that requires you to pass a tool / version and it resolves and fetches that version, so it doesn't quite cover what you're asking for.

Can you describe a bit more the use-case where you want to download the pinned version without running a command?

One potential option at the moment is you could run `node --version`, which will force the download but won't run anything substantial, just printing out the Node version (which should match what you have pinned).

**Comment by maraisr** [comment]
Yeah sure no worries; Yeah so where I really needed a feature like this was as follows.

- I use a popular JS library called huksy to mannage my git-hooks. One of my hooks run pre-commit using a node script.
- 5 commits ago I had node@8 pinned.
- HEAD now has node@12 pinned.
- I rebase before raising a PR.
- During rebase, I had a yarn warning about engines. As one of my deps wanted node@12, and had node@8 installed.
- I dare say that was because at HEAD I had node_modules with node@12 deps
- But during rebase, volta installed node@8 because of the pre-commit hook. But this didnt also install yarn deps at that point either.
- So git-rebase bails out, and gives me the ability to correct that. So at this point, I'd like to say, install the pinned copy of node (whatever that is), and run yarn install. So my hooks could run for that point.
- _or_ I could simply _assume_ the hooks are okay, and just `--continue`.

So yeah, tldr. running `yarn install` doesnt also trigger a volta node version check. So would be nice to install the toolchain as state of play, so peripheral tooling can operate.

I suppose, and like you said; the simplest way to get around this, would probably to just run a `node -v && yarn --version` as a pre-commit hook, before husky runs.

**Comment by charlespierce** [comment]
Let me restate to make sure I understand what's going on:

1. Your current `node_modules` has `node@12` dependencies installed
2. You rebase, and in one of the replayed commits, `package.json` had `node@8` (presumably a later commit then upgraded to Node 12).
3. Due to husky pre-commit, when applying the rebase, a Node script is run, at which point Volta checks, sees `node@8`, and uses that to execute the hook.
4. This causes errors because your deps were installed on `node@12`, so there is a compatibility mismatch with `node@8`.

If that's accurate, then I don't think the command you're asking for would change anything, because if you ran it before the rebase, you would fetch `node@12`, so you would be in the same situation. If you ran it as part of the pre-commit hook, you would get `node@8`, but you already get that from Volta during the hook, and that's ultimately the problem.

I think this specific issue is going to be a sticky spot with Volta, because Volta  re-evaluates the current Node version on every execution, but in this specific case of a rebase with pre-commit hooks, you actually want to tell Volta "No, just always use _this_ specific version". We don't currently have something like that, nor do we have something planned that would work with Husky. So it may be the case that you need to do some manual work when working with a rebase over a boundary with different Node versions, while you have a pre-commit hook, and native dependencies that care about Node version compatibility :smiley:

**Comment by charlespierce** [comment]
> So yeah, tldr. running yarn install doesnt also trigger a volta node version check.

Also wanted to point out that I don't believe this is true. Every command that goes through Volta re-evaluates the context on execution, so running `yarn install` _would_ result in Volta reading `package.json` to determine the currently pinned `node` and `yarn` versions and using those to run the command.

**Comment by maraisr** [comment]
Maybe I'll do more research, but dont think git-hooks run from the cwd.

Anyhow; I think what you said makes sense. Because really shouldn't should never need to install pinned copies, its kinda of supposed to be transparent.

Just as a small side note before you/I close this issue. Do you know of a way to get access to the `yarn` bin folder for use in IntelliJ? Somewhere in the `~/.volta` folder?

**Comment by charlespierce** [comment]
Oh the CWD thing could be another wrinkle, but I believe in my experience Volta works in git hooks, so I think it may work.

To your other question, I’m not sure about the IntelliJ interaction, but we do have a command `volta which yarn`, which will show the actual Yarn binary that would get executed if you ran `yarn`. Would that get what you need for IntelliJ?

**Comment by maraisr** [comment]
Ah! Yeah, that is leading me closer. I think Intellij Wants the folder that wraps the bin folder. Got absolutely no problem referencing node. Not sure if its the correct one from the workspace. But look thanks so much for your help! Think I'll close this issue for now.

But I am just wondering, wouldn't a `volta install pinned` help intellij users, that don't run `node` but rather need a reference to the binary, so it can spin up language servers and what not?

**Comment by charlespierce** [comment]
Yeah, a command like that may be useful; there's a higher-level investigation that we still need to do into IDE support in general, to ensure those things work cleanly with Volta. Ideally we'd have some way of specifying the CWD as well, so that the "correct" versions of Node / Yarn are used, but that may not be possible in all editors.

**Comment by charlespierce** [comment]
Another voice for the original request, @joshuarli raised the point that our auto-download can feel a bit too much like "magic", so giving the option to install pinned tools if they aren't already installed could give users some additional control over how they use Volta.

**Comment by joshuarli** [comment]
> Another voice for the original request, @joshuarli raised the point that our auto-download can feel a bit too much like "magic", so giving the option to install pinned tools if they aren't already installed could give users some additional control over how they use Volta.

Yeah. For the lazy: https://github.com/getsentry/sentry-docs/pull/1495#discussion_r377893906

**Comment by idleberg** [comment]
I'd like to add to this issue, if you don't mind. I'm also using husky and noticed that my git hooks are *always* run with the globally installed Node version, not the one pinned in the manifest. While I haven't had any problems with this, I would prefer if the pinned version was used - simply to rule out possible future problems related to the Node version.

**Comment by charlespierce** [comment]
@idleberg That's odd and I think perhaps a separate issue. We also use Node and I confirmed locally that it does use the local pinned version when I run `git commit`. How are you committing the files? Is it possibly through an IDE or something that isn't picking up the appropriate PATH settings to load Volta?

**Comment by idleberg** [comment]
> How are you committing the files? Is it possibly through an IDE or something that isn't picking up the appropriate PATH settings to load Volta?

It looks to me like a Windows-related issue, I've conducted a series of tests on my private computer (macOS) and my work computer (Windows 10):

- on macOS everything works as expected
- committing on Windows, from Microsoft Terminal and the VSCode terminal, works as expected (Git Bash and WSL bash)
- committing in PowerShell will run the git hooks using the globally installed Node version, both in Microsoft Terminal and the VSCode terminal

PS: Should I repeat my comment in a new issue?

**Comment by charlespierce** [comment]
@idleberg Thanks for narrowing it down! If you wouldn't mind, yeah could you create a new issue with what you've found so far? If you also have a few minutes, could you try running the commit in PowerShell after setting the environment variable `VOLTA_LOGLEVEL` to `debug`? It looks like in PowerShell the command `$env:VOLTA_LOGLEVEL = 'debug'` should set that value. Assuming things are properly going through Volta, that should output additional logging information about where it found the platform information, which should help debug further.

**Comment by idleberg** [comment]
One final note: when I conducted my tests to open a new issue, I could no longer produce the issues described in my previous posting. I suspect the problem was on my end, possibly a conflict with other Node version managers. So, don't expect a new issue unless I run into the problem again!
